maestrowf/interfaces/script/localpooladapter.py

```diff
@@ -0,0 +1,204 @@
+
+        :param step: The StudyStep instance this submission is based on.
+        :param path: Local path to the script to be executed.
+        :param cwd: Path to the current working directory.
+        :param job_map: A dictionary mapping step names to their job
+                        identifiers.
+        :param env: A dict containing a modified environment for execution.
+        :returns: The return status of the submission command and job
+                  identifier.
+        """
+        try:
+            job_id = self._pool.submit(path, cwd)
+            ret_code = 0
+            submit_status = SubmissionCode.OK
+        except Exception as exception:
+            LOGGER.error("Encountered exception: %s", str(exception))
+            job_id = -1
+            ret_code = -1
+            submit_status = SubmissionCode.ERROR
+
+        return SubmissionRecord(submit_status, ret_code, job_id)
+
+    def check_jobs(self, joblist):
+        """
+        For the given job list, query execution status.
+
+        This method uses the scontrol show job <jobid> command and does a
+        regex search for job information.
+
+        :param joblist: A list of job identifiers to be queried.
+        :returns: The return code of the status query, and a dictionary of job
+                  identifiers to their status.
+        """
+        LOGGER.debug("Joblist type -- %s", type(joblist))
+        LOGGER.debug("Joblist contents -- %s", joblist)
+        if not joblist:
+            LOGGER.debug("Empty job list specified.")
+            return JobStatusCode.OK, {}
+        if not isinstance(joblist, list):
+            LOGGER.debug("Specified parameter is not a list.")
+            if isinstance(joblist, int):
+                LOGGER.debug("Integer found.")
+                joblist = [joblist]
+            else:
+                LOGGER.debug("Unknown type. Returning an error.")
+                return JobStatusCode.ERROR, {}
+
+        status = {jid: State.UNKNOWN for jid in joblist}
+        try:
+            for job_id, state in self._pool.get_all_status():
+                if job_id in status:
+                    status[job_id] = self._state(state)
+            chk_status = JobStatusCode.OK
+
+        except Exception as exception:
+            LOGGER.error(str(exception))
+            chk_status = JobStatusCode.ERROR
+
+        return chk_status, status
+
+    def cancel_jobs(self, joblist):
+        """
+        For the given job list, cancel each job.
+
+        :param joblist: A list of job identifiers to be cancelled.
+        :returns: The return code to indicate if jobs were cancelled.
+        """
+        # If we don"t have any jobs to check, just return status OK.
+        if not joblist:
+            return CancelCode.OK
+
+        c_status = set()
+        for job in joblist:
+            c_status.add(self._pool.cancel(job))
+
+        c_return = CancelCode.OK
+        ret_code = 0
+        if ExecCancel.FAILED in c_status:
```

Comment:
Should this check all of them, i.e. counting the number of cancels that failed and ones that succeeded separately for more informative error messages?

Reply:
This design choice is a decision carried forward from SLURM (when I initially made this); in that perspective, the status is reported **if at least one** record reports the state. The issue with things like SLURM is that it's more efficient to do a bulk cancel, so there's no way to get a per job report out there (not without polling the scheduler to see if it took).